File: src/cs/IdpProtocol/IdpRouter.cs
// Copyright (c) VitalElement. All rights reserved.
// Licensed under the MIT license. See licence.md file in the project root for full license information.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;

namespace IdpProtocol
{
    public class IdpRouter : IdpNode, IAdaptorToRouterPort, IPacketTransmit
    {
        private List<IdpNode> _unenumeratedNodes;
        private Dictionary<UInt16, IdpNode> _enumeratedNodes;
        private Dictionary<UInt16, IdpAdaptor> _adaptors;
        private Dictionary<UInt16, UInt16> _routingTable;
        private UInt16 _nextAdaptorId;
        private IdpAdaptor _currentlyEnumeratingAdaptor;
        private int _lastAdaptorId = -1;

        public static readonly Guid RouterGuid = Guid.Parse("A1EE332D-5C7C-42FE-9519-54BDAC40CF21");

        public IdpRouter() : base(RouterGuid, "Network.Router")
        {
            _unenumeratedNodes = new List<IdpNode>();
            _enumeratedNodes = new Dictionary<UInt16, IdpNode>();
            _adaptors = new Dictionary<UInt16, IdpAdaptor>();
            _routingTable = new Dictionary<UInt16, UInt16>();

            Manager.RegisterCommand((UInt16)NodeCommand.RouterDetect, (i, o) =>
            {
                var address = i.Read<UInt16>();

                if (Address == UnassignedAddress)
                {
                    Address = address;
                    o.Write(true);
                    o.WithResponseCode(IdpResponseCode.OK);

                    SendRequest(i.Source, o);

                    return IdpResponseCode.Deferred;
                }
                else
                {
                    o.Write(false);
                    return IdpResponseCode.OK;
                }
            });

            Manager.RegisterCommand((UInt16)NodeCommand.RouterEnumerateNode, (i, o) =>
                                    {
                                        var address = i.Read<UInt16>();

                                        return OnEnumerateNodesCommand(i.Source, address, o);
                                    });

            Manager.RegisterCommand((UInt16)NodeCommand.RouterEnumerateAdaptor, (i, o) =>
                                    {
                                        var address = i.Read<UInt16>();
                                        var transactionId = i.Read<UInt32>();

                                        return OnEnumerateAdaptorCommand(i.Source, address, transactionId, o);
                                    });

            Manager.RegisterCommand((UInt16)NodeCommand.RouterPrepareToEnumerateAdaptors, (i, o) =>
            {
                foreach (var adaptor in _adaptors.Values)
                {
                    if (!adaptor.IsActive)
                    {
                        adaptor.IsRenumerated = false;
                    }
                }

                return IdpResponseCode.OK;
            });

            Manager.RegisterCommand((UInt16)NodeCommand.MarkAdaptorConnected, (i, o) =>
            {
                if (_currentlyEnumeratingAdaptor != null)
                {
                    _currentlyEnumeratingAdaptor.IsEnumerated = true;

                    _currentlyEnumeratingAdaptor = null;
                }
                else if(_lastAdaptorId != -1)
                {
                    _adaptors[(ushort)_lastAdaptorId].IsEnumerated = true;

                    _lastAdaptorId = -1;
                }

                return IdpResponseCode.OK;
            });

            TransmitEndpoint = this;

            _nextAdaptorId = 1;
        }

        /*public override async Task OnPollTimerTickAsync()
        {
            await base.OnPollTimerTickAsync();

            foreach(var adaptorEntry in _adaptors.Where(x=>x.Value.IsEnumerated))
            {
                var adaptor = adaptorEntry.Value;

                var outgoingTransaction = OutgoingTransaction.Create((UInt16)NodeCommand.RouterPoll, CreateTransactionId());

                var result = adaptor.Transmit(outgoingTransaction.ToPacket(Address, RouterPollAddress));

                if(result)
                {
                    var response = await Manager.WaitForResponseAsync(outgoingTransaction.TransactionId, 2500);

                    if(response == null || response.ResponseCode != IdpResponseCode.OK)
                    {
                        adaptor.IsEnumerated = false;
                    }
                }
                else
                {
                    adaptor.IsEnumerated = false;
                }
            }
        }*/

        protected override void OnReset()
        {
            foreach (var node in _enumeratedNodes.Values.ToList())
            {
                if (node.Address != 0x0001)
                {
                    MarkUnenumerated(node);
                }
            }

            foreach (var adaptor in _adaptors.Values)
            {
                adaptor.IsEnumerated = false;
            }

            base.OnReset();
        }

        private IdpNode FindNode(UInt16 address)
        {
            if (_enumeratedNodes.ContainsKey(address))
            {
                return _enumeratedNodes[address];
            }

            return null;
        }

        private IdpResponseCode OnEnumerateNodesCommand(UInt16 source, UInt16 address, OutgoingTransaction outgoing)
        {
            if (_unenumeratedNodes.Count != 0)
            {
                var node = _unenumeratedNodes.First();

                node.Address = address;
                //node.SetEnumerated();

                if (MarkEnumerated(node))
                {
                    outgoing.Write(true);
                    outgoing.WithResponseCode(IdpResponseCode.OK);

                    SendRequest(address, source, outgoing);

                    return IdpResponseCode.Deferred;
                }
                else
                {
                    node.Address = UnassignedAddress;

                    outgoing.Write(false);

                    return IdpResponseCode.InvalidParameters;
                }
            }
            else
            {
                outgoing.Write(false);
            }

            return IdpResponseCode.OK;
        }

        private IdpResponseCode OnEnumerateAdaptorCommand(UInt16 source, UInt16 address, UInt32 transactionId, OutgoingTransaction outgoing)
        {
            var adaptor = GetNextUnenumeratedAdaptor();

            if (adaptor == null)
            {
                outgoing.Write(false);

                return IdpResponseCode.OK;
            }
            else
            {
                adaptor.IsRenumerated = true;

                bool adaptorProbed = false;

                if (!adaptor.IsEnumerated)
                {
                    var request = new OutgoingTransaction((UInt16)NodeCommand.RouterDetect, transactionId, IdpCommandFlags.None).Write(address);

                    adaptorProbed = SendRequest(adaptor, source, 0xFFFF, request);

                    if (adaptorProbed)
                    {
                        _currentlyEnumeratingAdaptor = adaptor;
                    }
                }

                outgoing.Write(true).Write(adaptorProbed).WithResponseCode(IdpResponseCode.OK);

                SendRequest(source, outgoing);

                return IdpResponseCode.Deferred;
            }
        }

        private bool SendRequest(IPacketTransmit adaptor, UInt16 source, UInt16 destination, OutgoingTransaction request)
        {
            return adaptor.Transmit(request.ToPacket(source, destination));
        }

        private IdpAdaptor GetNextUnenumeratedAdaptor()
        {
            return _adaptors.Values.FirstOrDefault(a => !a.IsEnumerated);
        }

        private IdpAdaptor GetNextReUnenumeratedAdaptor()
        {
            return _adaptors.Values.FirstOrDefault(a => !a.IsRenumerated);
        }

        public bool Transmit(ushort adaptorId, IdpPacket packet)
        {
            var source = packet.Source;

            if (source != UnassignedAddress && adaptorId != 0xFFFF)
            {
                //if (!_adaptors[adaptorId].IsEnumerated)
                {
                    _lastAdaptorId = adaptorId;
                }

                if (!(source == 1 && _routingTable.ContainsKey(source)))
                {
                    _routingTable[source] = adaptorId;
                }
            }

            return Route(packet);
        }

        public bool MarkEnumerated(IdpNode node)
        {
            bool result = false;

            if (FindNode(node.Address) == null)
            {
                _unenumeratedNodes.Remove(node);

                _enumeratedNodes[node.Address] = node;

                result = true;
            }

            return result;
        }

        public void MarkUnenumerated(IdpNode node)
        {
            if (!_unenumeratedNodes.Contains(node))
            {
                _unenumeratedNodes.Add(node);
            }

            if (_enumeratedNodes.ContainsKey(node.Address))
            {
                _enumeratedNodes.Remove(node.Address);
            }
        }

        public bool AddNode(IdpNode node)
        {
            bool result = false;

            if (node.Address == UnassignedAddress)
            {
                _unenumeratedNodes.Add(node);
                result = true;
            }
            else
            {
                if (FindNode(node.Address) == null)
                {
                    _enumeratedNodes[node.Address] = node;
                    result = true;
                }
            }

            if (result)
            {
                node.TransmitEndpoint = this;
                node.Enabled = true;
            }

            return result;
        }

        public void RemoveNode(IdpNode node)
        {
            if (node.Address == UnassignedAddress)
            {
                _unenumeratedNodes.Remove(node);
            }
            else if (FindNode(node.Address) != null)
            {
                _enumeratedNodes.Remove(node.Address);
            }

            node.Address = UnassignedAddress;
        }

        public bool AddAdaptor(IdpAdaptor adaptor)
        {
            bool result = false;

            adaptor.LocalPort = this;

            adaptor.Id = _nextAdaptorId;

            _adaptors[_nextAdaptorId] = adaptor;

            _nextAdaptorId++;

            return result;
        }

        public bool Route(IdpPacket packet)
        {
            var destination = packet.Destination;
            var source = packet.Source;

            packet.ResetReadToPayload();
            var command = packet.Read<UInt16>();

            packet.ResetRead();

            //Debug.WriteLine($"Router: 0x{Address.ToString("X4")} from: 0x{source.ToString("X4")}, to: 0x{ destination.ToString("X4")} Command: 0x{ command.ToString("X4")} ({(NodeCommand)command})");

            if (destination == 0 && Address != UnassignedAddress)
            {
                foreach (var adaptor in _adaptors.Values)
                {
                    if (!_routingTable.ContainsKey(source) || adaptor != _adaptors[_routingTable[source]])
                    {
                        packet.ResetRead();

                        adaptor.Transmit(packet);
                    }
                }

                foreach (var node in _enumeratedNodes.Values)
                {
                    packet.ResetRead();

                    var packetResponse = node.ProcessPacket(packet);

                    if (packetResponse != null)
                    {
                        Route(packetResponse);
                    }
                }


                packet.ResetRead();

                var response = ProcessPacket(packet);

                if (response != null)
                {
                    Route(response);
                }

                return true;
            }
            else if (destination == RouterPollAddress && Address != UnassignedAddress)
            {
                var response = ProcessPacket(packet);

                if(response != null)
                {
                    return Route(response);
                }

                return false;
            }
            else
            {
                IdpNode node = null;

                if (destination == Address)
                {
                    node = this;
                }
                else
                {
                    node = FindNode(destination);
                }

                if (node != null)
                {
                    var responsePacket = node.ProcessPacket(packet);

                    if (responsePacket != null)
                    {
                        return Route(responsePacket);
                    }

                    return true;
                }
                else
                {
                    if (_routingTable.ContainsKey(destination))
                    {
                        return _adaptors[_routingTable[destination]].Transmit(packet);
                    }
                    else
                    {
                        if (destination == 0xFFFF && Address != 0xFFFF)
                        {
                            return false;
                        }

                        return _adaptors.FirstOrDefault().Value.Transmit(packet);
                    }
                }
            }
        }

        public bool Transmit(IdpPacket packet)
        {
            return Transmit(0xFFFF, packet);
        }
    }
}

File: src/cpp/IdpRouter.cpp
// Copyright (c) VitalElement. All rights reserved.
// Licensed under the MIT license. See licence.md file in the project root for
// full license information.
#include "IdpRouter.h"
#include "Trace.h"
#include <algorithm>

static constexpr uint16_t AdaptorNone = 0xFFFF;

IdpRouter::IdpRouter () : IdpNode (RouterGuid, "Network.Router")
{
    _currentlyEnumeratingAdaptor = nullptr;

    Manager ().RegisterCommand (
        static_cast<uint16_t> (NodeCommand::RouterDetect),
        [&](std::shared_ptr<IncomingTransaction> incoming,
            std::shared_ptr<OutgoingTransaction> outgoing) {
            auto address = incoming->Read<uint16_t> ();

            if (this->Address () == UnassignedAddress)
            {
                this->Address (address);
                outgoing->Write (true);
                outgoing->WithResponseCode (IdpResponseCode::OK);

                IdpNode::SendRequest (incoming->Source (), outgoing);

                return ::IdpResponseCode::Deferred;
            }
            else
            {
                outgoing->Write (false);
                return IdpResponseCode::OK;
            }
        });

    Manager ().RegisterCommand (
        static_cast<uint16_t> (NodeCommand::RouterEnumerateNode),
        [&](std::shared_ptr<IncomingTransaction> incoming,
            std::shared_ptr<OutgoingTransaction> outgoing) {
            auto address = incoming->Read<uint16_t> ();

            return this->HandleEnumerateNodesCommand (incoming->Source (),
                                                      address, outgoing);
        });

    Manager ().RegisterCommand (
        static_cast<uint16_t> (NodeCommand::RouterEnumerateAdaptor),
        [&](std::shared_ptr<IncomingTransaction> incoming,
            std::shared_ptr<OutgoingTransaction> outgoing) {
            auto address = incoming->Read<uint16_t> ();
            auto transactionId = incoming->Read<uint32_t> ();

            return this->HandleEnumerateAdaptorCommand (
                incoming->Source (), address, transactionId, outgoing);
        });

    Manager ().RegisterCommand (
        static_cast<uint16_t> (NodeCommand::RouterPrepareToEnumerateAdaptors),
        [&](std::shared_ptr<IncomingTransaction> incoming,
            std::shared_ptr<OutgoingTransaction> outgoing) {
            auto it = _adaptors.begin ();

            while (it != _adaptors.end ())
            {
                it->second->IsReEnumerated (false);

                it++;
            }
            return IdpResponseCode::OK;
        });

    Manager ().RegisterCommand (
        static_cast<uint16_t> (NodeCommand::MarkAdaptorConnected),
        [&](std::shared_ptr<IncomingTransaction> incoming,
            std::shared_ptr<OutgoingTransaction> outgoing) {
            Trace::WriteLine ("Marking adaptor enumerated");

            if (_currentlyEnumeratingAdaptor != nullptr)
            {
                _currentlyEnumeratingAdaptor->IsEnumerated (true);

                _currentlyEnumeratingAdaptor = nullptr;
            }

            return IdpResponseCode::OK;
        });


    // Routers endpoint is themselves, routing tables will find the correct
    // adaptor.
    TransmitEndpoint (*this);

    _nextAdaptorId = 1;
}

IdpRouter::~IdpRouter ()
{
}

void IdpRouter::OnReset ()
{
    auto it = _enumeratedNodes.begin ();

    while (it != _enumeratedNodes.end ())
    {
        if (it->second->Address () != 0x0001)
        {
            MarkUnenumerated (*it->second);

            it = _enumeratedNodes.erase (it);
        }
        else
        {
            it++;
        }
    }


    auto it1 = _adaptors.begin ();

    while (it1 != _adaptors.end ())
    {
        it1->second->IsEnumerated (false);

        it1++;
    }

    IdpNode::OnReset ();
}

bool IdpRouter::SendRequest (IPacketTransmit& adaptor, uint16_t source,
                             uint16_t destination,
                             std::shared_ptr<OutgoingTransaction> request)
{
    return adaptor.Transmit (request->ToPacket (source, destination));
}

IdpNode* IdpRouter::FindNode (uint16_t address)
{
    auto it = _enumeratedNodes.find (address);

    if (it != _enumeratedNodes.end ())
    {
        return _enumeratedNodes[address];
    }

    return nullptr;
}

bool IdpRouter::AddAdaptor (IAdaptor& adaptor)
{
    adaptor.SetLocal (*this);

    adaptor.AdaptorId (_nextAdaptorId); // TODO assign unique id.

    _adaptors[_nextAdaptorId] = &adaptor;

    _nextAdaptorId++;

    return true;
}

bool IdpRouter::AddNode (IdpNode& node)
{
    bool result = false;

    if (node.Address () == UnassignedAddress)
    {
        _unenumeratedNodes.push_front (&node);
        result = true;
    }
    else
    {
        auto existing = FindNode (node.Address ());

        if (existing == nullptr)
        {
            _enumeratedNodes[node.Address ()] = &node;

            result = true;
        }
    }

    if (result)
    {
        node.TransmitEndpoint (*this);
    }

    node.Enabled (true);

    return result;
}

void IdpRouter::RemoveNode (IdpNode& node)
{
    if (node.Address () == UnassignedAddress)
    {
        _unenumeratedNodes.remove (&node);
    }
    else if (FindNode (node.Address ()) != nullptr)
    {
        auto it = _enumeratedNodes.find (node.Address ());

        if (it != _enumeratedNodes.end ())
        {
            _enumeratedNodes.erase (it);
        }
    }

    node.Address (UnassignedAddress);
}

bool IdpRouter::MarkEnumerated (IdpNode& node)
{
    bool result = false;

    if (FindNode (node.Address ()) == nullptr)
    {
        auto it = std::find (_unenumeratedNodes.begin (),
                             _unenumeratedNodes.end (), &node);

        if (it != _unenumeratedNodes.end ())
        {
            _unenumeratedNodes.erase (it);
        }

        _enumeratedNodes[node.Address ()] = &node;

        result = true;
    }

    return result;
}

void IdpRouter::MarkUnenumerated (IdpNode& node)
{
    auto it = std::find (_unenumeratedNodes.begin (), _unenumeratedNodes.end (),
                         &node);

    if (it == _unenumeratedNodes.end ())
    {
        _unenumeratedNodes.push_front (&node);
    }
}

bool IdpRouter::Transmit (std::shared_ptr<IdpPacket> packet)
{
    Transmit (AdaptorNone, packet);

    return true;
}

IdpResponseCode IdpRouter::HandleEnumerateNodesCommand (
    uint16_t source, uint16_t address,
    std::shared_ptr<OutgoingTransaction> outgoing)
{
    if (!_unenumeratedNodes.empty ())
    {
        auto& node = *_unenumeratedNodes.back ();

        node.Address (address);

        if (MarkEnumerated (node))
        {
            outgoing->Write (true);
            outgoing->WithResponseCode (IdpResponseCode::OK);

            IdpNode::SendRequest (address, source, outgoing);

            return IdpResponseCode::Deferred;
        }
        else
        {
            node.Address (UnassignedAddress);

            outgoing->Write (false);

            return IdpResponseCode::InvalidParameters;
        }
    }
    else
    {
        outgoing->Write (false);
    }

    return IdpResponseCode::OK;
}

IdpResponseCode IdpRouter::HandleEnumerateAdaptorCommand (
    uint16_t source, uint16_t address, uint32_t transcationId,
    std::shared_ptr<OutgoingTransaction> outgoing)
{
    auto adaptor = GetNextUnenumeratedAdaptor (true);

    if (adaptor == nullptr)
    {
        Trace::WriteLine ("No Adaptor found on Router: %u", "IdpRouter",
                          Address ());
        outgoing->Write (false);

        return IdpResponseCode::OK;
    }
    else
    {
        adaptor->IsReEnumerated (true);

        auto outgoingTransaction =
            OutgoingTransaction::Create (
                static_cast<uint16_t> (NodeCommand::RouterDetect),
                transcationId, IdpCommandFlags::None)
                ->Write (address);

        bool adaptorSent = false;

        if (!adaptor->IsEnumerated ())
        {
            adaptorSent =
                SendRequest (*adaptor, source, 0xFFFF, outgoingTransaction);

            if (adaptorSent)
            {
                _currentlyEnumeratingAdaptor = adaptor;
            }
        }

        Trace::WriteLine ("Adaptor (%s) found on Router: %u, active: %s",
                          "IdpRouter", adaptor->Name (), Address (),
                          adaptor->IsEnumerated () ? "true" : "false");

        outgoing->Write (true)
            ->Write (adaptorSent)
            ->WithResponseCode (IdpResponseCode::OK);

        IdpNode::SendRequest (source, outgoing);

        return IdpResponseCode::Deferred;
    }
}

IAdaptor* IdpRouter::GetNextUnenumeratedAdaptor (bool reenumeration)
{
    auto it = _adaptors.begin ();

    while (it != _adaptors.end ())
    {
        if (reenumeration && !it->second->IsReEnumerated ())
        {
            break;
        }

        if (!reenumeration && !it->second->IsEnumerated ())
        {
            break;
        }

        it++;
    }

    if (it != _adaptors.end ())
    {
        return it->second;
    }

    return nullptr;
}

void IdpRouter::Transmit (uint16_t adaptorId, std::shared_ptr<IdpPacket> packet)
{
    auto source = packet->Source ();

    if (source != UnassignedAddress && adaptorId != 0xFFFF)
    {
        _adaptors[adaptorId]->IsEnumerated (true);

        if (!(source == 1 &&
              _routingTable.find (source) != _routingTable.end ()))
        {
            _routingTable[source] =
                adaptorId; // replace any existing route with
                           // the one the packet just came from.
        }
    }

    Route (packet);
}

void IdpRouter::Route (std::shared_ptr<IdpPacket> packet)
{
    auto destination = packet->Destination ();
    auto source = packet->Source ();

    packet->ResetReadToPayload ();
    auto command = packet->Read<uint16_t> ();
    auto transactionId = packet->Read<uint32_t> ();
    packet->Read<uint8_t> ();

    if (command == (uint16_t) NodeCommand::Response)
    {
        packet->Read<uint8_t> ();
        auto response = packet->Read<uint16_t> ();

        Trace::Write (
            "R:0x%04x S:0x%04x D:0x%04x "
            "C:0x%04x <%s ",
            "IdpRouter", Address (), source, destination, command,
            IdpNode::GetNodeCommandDescription ((NodeCommand) response));
    }
    else
    {
        Trace::Write (
            "R:0x%04x S:0x%04x D:0x%04x "
            "C:0x%04x >%s ",
            "IdpRouter", Address (), source, destination, command,
            IdpNode::GetNodeCommandDescription ((NodeCommand) command));
    }

    if (packet->Length () <= 32)
    {
        Trace::AppendBuffer (packet->Data (), packet->Length ());
    }

    Trace::AppendLine ("");

    packet->ResetRead ();

    if (destination == 0)
    {
        auto ait = _adaptors.begin ();

        auto receivedOn = _routingTable.find (source);

        while (ait != _adaptors.end ())
        {
            if (receivedOn == _routingTable.end () ||
                ait->second != _adaptors[receivedOn->second])
            {
                packet->ResetRead ();

                ait->second->Transmit (packet);
            }

            ait++;
        }

        auto it = _enumeratedNodes.begin ();

        while (it != _enumeratedNodes.end ())
        {
            packet->ResetRead ();

            auto response = it->second->ProcessPacket (packet);

            if (response != nullptr)
            {
                Route (response);
            }

            it++;
        }

        packet->ResetRead ();

        auto response = ProcessPacket (packet);

        if (response != nullptr)
        {
            Route (response);
        }
    }
    else
    {
        IdpNode* node = nullptr;

        if (destination == Address ())
        {
            node = this;
        }
        else
        {
            node = FindNode (packet->Destination ());
        }

        if (node != nullptr)
        {
            auto responsePacket = node->ProcessPacket (packet);

            if (responsePacket != nullptr)
            {
                Route (responsePacket);
            }
        }
        else
        {
            // todo find adaptor with that node.
            auto it = _routingTable.find (destination);

            if (it != _routingTable.end ())
            {
                _adaptors[it->second]->Transmit (packet);
            }
            else if (destination != UnassignedAddress)
            {
                // Probably safe to assume that we should transmit in same route
                // as master?
                Trace::WriteLine ("Packet Dropped: Unknown Route", "IdpRouter");
            }
        }
    }
}
